clang/lib/AST/StmtProfile.cpp

```diff
@@ -1333,7 +1333,13 @@ void StmtProfiler::VisitPredefinedExpr(const PredefinedExpr *S) {
 void StmtProfiler::VisitIntegerLiteral(const IntegerLiteral *S) {
   VisitExpr(S);
   S->getValue().Profile(ID);
-  ID.AddInteger(S->getType()->castAs<BuiltinType>()->getKind());
+
+  QualType T = S->getType();
+  ID.AddInteger(T->getTypeClass());
```

Comment:
This addition of the type class doesn't observe the `Canonical` attribute, leading to IntegerLiterals with different types but same canonical types mismatching even when `Canonical` is true.

Reply:
Oooh, good catch!  Thanks for that!  Can you contribute a patch to fix that? (That is, canonicalize T).

Reply:
Sure thing. I haven't been able to find a reproducer for upstream clang, so I don't have a test case, but it doesn't seem to break any tests at least. Waiting for it to push to github now.